docs/design/2024-05-10-extension-authentication-plugin.md

```diff
@@ -0,0 +1,231 @@
++ `session.Auth` -> `privileges.ConnectionVerification`
+  + Authenticate the user based on the plugin specified. For password based authentication such as `mysql_native_password`, TiDB will compare the password digest with the locally stored hash
+
+
+Authorization is more straight-forward (more details available in the [official developer guide](https://pingcap.github.io/tidb-dev-guide/understand-tidb/dql.html)):
++ `server.onConn`
++ `server.HandleStmt`
++ `server.ExecuteStmt`
++ `executor.Compile`
++ `Planner.optimize`
++ `privileges.RequestVerification`
+  + Note that this function will be called multiple times, one time for each privilege required by the SQL statements
+
+## Detailed Design
+
+We propose improving the extension system to support the following functionalities:
++ Allow user defined extra authentication check to be executed during connection establishment phase
+  + Current `SessionHandler` [support](https://github.com/pingcap/tidb/tree/master/pkg/extension) in the extension system only allows running observer functions that do not affect the connection attempt result
+  + Errors returned from this check will cause the connect attempt to fail
++ Allow user defined extra SQL privilege check to be executed before SQL statements are executed
+
+Note that the above plugin SQL privilege checks are only additional to the existing checks done by TiDB. If the existing checks done by TiDB fails, the plugin checks won’t be invoked. This means that the user should already have the corresponding SQL privileges granted to itself before we can trigger the plugin privilege check, otherwise the SQL request should be rejected beforehand since the user is missing the privileges.
+
+Users should be able to develop custom auth plugins which handle the above two checks. For example, if a user develops a plugin called `authentication_my_plugin`, then once the plugin is loaded, the user can run the following SQL statement:
+
+```sql
+CREATE USER 'my_user'@'%' IDENTIFIED WITH 'authentication_my_plugin' AS 'optional_authentication_pwd';
+```
+
+After this, `my_user`’s connection verification and privilege verification will be handled by the plugin in addition to the existing default TiDB checks. Users not identified with the plugin will not go through the plugin authorization and authentication checks.
+
+### Interface
+We propose a set of interfaces for users to implement an auth plugin that are similar to what is offered in MySQL’s authentication plugins. We will modify TiDB’s code to call these interfaces during the authentication/authorization flows. Most of these interfaces will receive parameters that are used in the existing auth flows in TiDB, so that the plugin developer can access as much information as possible.
+
+Following is the interface for an auth plugin:
+```go
+// AuthPlugin is an interface for authentication plugins.
+type AuthPlugin interface {
```

Comment:
Good idea. Will make this a struct instead.

One issue is that user might not implement all the functions required, which could cause NPE when invoking them. When we register the plugin in the manifest, we can check whether all required functions are not `nil`, otherwise return error and fail the start up. What do you think?